File: rock/sdk/sandbox/oss_client.py
"""OssClient — encapsulates all OSS interactions for a Sandbox.

Holds OSS state (bucket, token expiration, async persistence tasks) and
exposes upload / download / persistence operations. Composed by Sandbox.
"""

from __future__ import annotations

import asyncio
import hashlib
import re
import shlex
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import TYPE_CHECKING

import oss2

from rock.actions.sandbox.request import Command
from rock.actions.sandbox.response import DownloadFileResponse, UploadResponse
from rock.logger import init_logger
from rock.utils.http import HttpUtils

if TYPE_CHECKING:
    from rock.sdk.sandbox.client import Sandbox

logger = init_logger(__name__)


@dataclass
class OssClientConfig:
    """Resolved OSS configuration from admin /get_token response."""

    endpoint: str
    bucket: str
    region: str
    prefix: str = ""  # Transfer-object key prefix from server response (e.g. "rock-transfer/")


class OssClient:
    """OSS operations for a single Sandbox instance."""

    def __init__(self, sandbox: Sandbox):
        self._sandbox = sandbox
        self._bucket = None
        self._token_expire_time: str | None = None
        self._client_config: OssClientConfig | None = None
        self._pending_persistence_tasks: set[asyncio.Task] = set()

    @staticmethod
    def _compute_object_name(
        sandbox_id: str,
        local_path: str,
        sandbox_path: str,
        prefix: str | None = None,  # NEW: server-pushed transfer prefix (e.g. "rock-transfer/")
    ) -> str:
        # Prefer sandbox basename: the OSS object mirrors a sandbox-side file,
        # so naming it after the sandbox path keeps OSS-side names meaningful
        # even when local destinations differ (e.g. download to a renamed file).
        payload = f"{sandbox_id}|{local_path}|{sandbox_path}"
        digest = hashlib.sha256(payload.encode("utf-8")).hexdigest()
        filename = Path(sandbox_path).name or Path(local_path).name
        # Honor server-pushed Prefix so chatos-rock's rock-transfer/ lifecycle catches the object.
        # Sanitize: trim whitespace, strip leading/trailing slashes, collapse internal "//" runs.
        # Without this, a stray prefix like " ", "/rock-transfer/" or "rock//transfer" would
        # produce a malformed OSS key (e.g. "   /file", "//file", "rock//transfer/file").
        clean_prefix = re.sub(r"/+", "/", (prefix or "").strip().strip("/"))
        clean_filename = f"{digest}-{filename}".lstrip("/")
        if clean_prefix:
            return f"{clean_prefix}/{clean_filename}"
        return clean_filename

    @staticmethod
    def _resolve_config(sts_response: dict) -> OssClientConfig | None:
        # Server-only: admin /get_token must return a complete OSS config.
        # If it doesn't, OSS is simply unavailable (no env fallback).
        resp_endpoint = sts_response.get("Endpoint")
        resp_bucket = sts_response.get("Bucket")
        resp_region = sts_response.get("Region")
        if resp_endpoint and resp_bucket and resp_region:
            return OssClientConfig(
                endpoint=resp_endpoint,
                bucket=resp_bucket,
                region=resp_region,
                prefix=sts_response.get("Prefix") or "",
            )

        # OSS unavailable: server did not supply a complete config.
        return None

    async def _get_sts_credentials(self) -> dict:
        """Fetch STS credentials and OSS config from /get_token endpoint.

        Returns the entire response result dict, which may include:
        - STS creds: AccessKeyId, AccessKeySecret, SecurityToken, Expiration
        - OSS config (if server is new + configured): Endpoint, Bucket, Region

        Side effect: caches Expiration in self._token_expire_time.
        """
        url = f"{self._sandbox._url}/get_token?account=primary"
        headers = self._sandbox._build_headers()
        response = await HttpUtils.get(url, headers)
        if response["status"] != "Success":
            raise Exception(f"Failed to get OSS STS token: {response.get('message', 'Unknown error')}")
        credentials = response["result"]
        self._token_expire_time = credentials["Expiration"]
        return credentials

    def _is_token_expired(self) -> bool:
        """Whether cached token is missing, malformed, or within 5min of expiration."""
        try:
            expire_time = datetime.fromisoformat(self._token_expire_time.replace("Z", "+00:00"))
            current_time = datetime.now(timezone.utc)
            effective_expire_time = expire_time - timedelta(minutes=5)
            return current_time >= effective_expire_time
        except (ValueError, AttributeError):
            return True

    @property
    def is_available(self) -> bool:
        """Whether OSS is available: bucket has been successfully initialized."""
        return self._bucket is not None

    async def ensure_setup(self) -> bool:
        """Ensure OSS bucket is set up and token is fresh. Idempotent.

        Returns True if OSS is available, False otherwise.
        """
        if self._bucket is not None and not self._is_token_expired():
            return True
        return await self._setup()

    async def _setup(self) -> bool:
        try:
            sts_response = await self._get_sts_credentials()
        except Exception as e:
            logger.warning(f"Failed to get STS credentials: {e}")
            return False

        config = self._resolve_config(sts_response)
        if config is None:
            return False

        try:
            auth = oss2.StsAuth(
                sts_response["AccessKeyId"],
                sts_response["AccessKeySecret"],
                sts_response["SecurityToken"],
            )
            self._bucket = oss2.Bucket(
                auth=auth,
                endpoint=config.endpoint,
                bucket_name=config.bucket,
                region=config.region,
            )
            self._client_config = config
            return True
        except Exception as e:
            logger.warning(f"Failed to initialize OSS bucket: {e}")
            self._bucket = None
            return False

    async def upload_via_oss(self, file_path: str, target_path: str) -> UploadResponse:
        """Upload a local file to sandbox via OSS as intermediary (large file path).

        Steps:
            1. Compute deterministic OSS object name
            2. Resumable upload local file to OSS
            3. Sign a temporary GET URL
            4. Inside sandbox: mkdir -p parent + wget the signed URL
            5. Verify target file exists in sandbox
        """
        from rock.sdk.sandbox.client import RunMode  # late import to avoid circular

        if self._bucket is None:
            return UploadResponse(success=False, message="OSS bucket not set up")

        file_name = Path(file_path).name
        oss_object_name = self._compute_object_name(
            sandbox_id=self._sandbox.sandbox_id,
            local_path=file_path,
            sandbox_path=target_path,
            prefix=self._client_config.prefix if self._client_config else None,  # NEW
        )

        try:
            oss2.resumable_upload(self._bucket, oss_object_name, file_path)
            url = self._bucket.sign_url("GET", oss_object_name, 600, slash_safe=True)

            # mkdir -p target parent (wget -O does not auto-create dirs in NOHUP mode)
            parent_dir = str(Path(target_path).parent)
            await self._sandbox.arun(
                cmd=f"mkdir -p {shlex.quote(parent_dir)}",
                wait_timeout=10,
                mode=RunMode.NORMAL,
            )

            # wget the signed URL.
            # Note: NO `-c` (continue/resume). With `-c`, wget skips download
            # entirely if the local file already exists with size matching the
            # remote object — but the OSS object name is derived from
            # (sandbox_id|local_path|sandbox_path), not file content, so a
            # repeat upload to the same target re-uses the same OSS key. The
            # remote object IS overwritten by `resumable_upload` above (new
            # content), but `wget -c` would compare sizes, see they match, and
            # exit 0 without fetching, leaving the sandbox file at its old
            # content while we still report success. `-O` alone forces wget
            # to truncate and rewrite, regardless of existing local state.
            download_cmd = f"wget -O {shlex.quote(target_path)} '{url}'"
            await self._sandbox.arun(cmd=download_cmd, wait_timeout=600, mode=RunMode.NOHUP)

            # Verify target exists in sandbox. Use execute() instead of arun(),
            # because arun() raises on non-zero exit codes; here exit_code=1
            # (file missing) is a normal branch we need to inspect.
            check = await self._sandbox.execute(Command(command=["test", "-f", target_path]))
            if check.exit_code != 0:
                return UploadResponse(
                    success=False,
                    message=f"Failed to upload file {file_name}, sandbox download phase failed",
                )
            return UploadResponse(
                success=True,
                message=f"Successfully uploaded file {file_name} to {target_path}",
            )
        except Exception as e:
            logger.warning(f"upload_via_oss failed: {e}")
            return UploadResponse(
                success=False,
                message=f"Failed to upload file {file_name} to {target_path}: {e}",
            )

    async def download_via_oss(self, remote_path: str, local_path: Path) -> DownloadFileResponse:
        """Download file from sandbox to local via OSS as intermediary.

        Note: ensure_setup must succeed before calling. Caller is LinuxFileSystem,
        which holds an `ensure_ossutil` helper for installing ossutil in the sandbox.
        """
        from rock.sdk.sandbox.client import RunMode  # late import

        if self._bucket is None or self._client_config is None:
            return DownloadFileResponse(success=False, message="OSS is not available")

        # Verify source file exists in sandbox (must be regular file).
        # Use execute() instead of arun(): arun() raises on non-zero exit
        # codes, but exit_code=1 (file missing) is the branch we want to act on.
        check = await self._sandbox.execute(Command(command=["test", "-f", remote_path]))
        if check.exit_code != 0:
            return DownloadFileResponse(
                success=False,
                message=(
                    f"Source file not found or is not a regular file in sandbox: {remote_path}. "
                    "Note: Only regular files are supported. For directories, create a tar archive first."
                ),
            )

        # Caller (LinuxFileSystem) is responsible for ensure_ossutil before calling here.

        # Refresh STS creds for ossutil (use existing helper, will refresh if expired)
        if self._is_token_expired():
            await self._setup()
        sts_response = await self._get_sts_credentials()
        access_key_id = sts_response["AccessKeyId"]
        access_key_secret = sts_response["AccessKeySecret"]
        security_token = sts_response["SecurityToken"]

        # Upload sandbox file to OSS via ossutil
        oss_object_name = self._compute_object_name(
            sandbox_id=self._sandbox.sandbox_id,
            local_path=str(local_path),
            sandbox_path=remote_path,
            prefix=self._client_config.prefix if self._client_config else None,
        )
        oss_url = f"oss://{self._client_config.bucket}/{oss_object_name}"

        ossutil_inner = (
            f"ossutil cp {shlex.quote(remote_path)} {shlex.quote(oss_url)}"
            f" --access-key-id {shlex.quote(access_key_id)}"
            f" --access-key-secret {shlex.quote(access_key_secret)}"
            f" --sts-token {shlex.quote(security_token)}"
            f" --endpoint {shlex.quote(self._client_config.endpoint)}"
            f" --region {shlex.quote(self._client_config.region)}"
        )
        upload_cmd = f"bash -c {shlex.quote(ossutil_inner)}"
        upload_resp = await self._sandbox.arun(cmd=upload_cmd, mode=RunMode.NOHUP)
        if upload_resp.exit_code != 0:
            return DownloadFileResponse(
                success=False,
                message=f"Failed to upload file to OSS (exit_code={upload_resp.exit_code}): {upload_resp.output}",
            )

        # Download from OSS to local via oss2
        auth = oss2.StsAuth(access_key_id, access_key_secret, security_token)
        bucket = oss2.Bucket(
            auth, self._client_config.endpoint, self._client_config.bucket, region=self._client_config.region
        )
        local = Path(local_path).expanduser().resolve()
        local.parent.mkdir(parents=True, exist_ok=True)
        try:
            bucket.get_object_to_file(oss_object_name, str(local))
        except Exception as e:
            return DownloadFileResponse(success=False, message=f"Failed to download from OSS: {e}")

        if not local.exists():
            return DownloadFileResponse(success=False, message=f"Downloaded file not found at: {local}")

        return DownloadFileResponse(success=True, message=f"Successfully downloaded {remote_path} to {local}")

    async def schedule_async_persistence(self, local_path: str, sandbox_path: str) -> str | None:
        """Fire-and-forget: upload local file to OSS in the background.

        Returns the OSS object key if scheduled, None if OSS is unavailable.
        Failures are logged as warnings; main flow is unaffected.
        """
        if self._bucket is None:
            return None

        oss_object_name = self._compute_object_name(
            sandbox_id=self._sandbox.sandbox_id,
            local_path=local_path,
            sandbox_path=sandbox_path,
            prefix=self._client_config.prefix if self._client_config else None,
        )
        task = asyncio.create_task(self._persist_to_oss(local_path, oss_object_name))
        self._pending_persistence_tasks.add(task)
        task.add_done_callback(self._pending_persistence_tasks.discard)
        return oss_object_name

    async def _persist_to_oss(self, local_path: str, oss_object_name: str) -> None:
        try:
            await asyncio.to_thread(oss2.resumable_upload, self._bucket, oss_object_name, local_path)
            logger.info(f"OSS persisted: {oss_object_name}")
        except Exception as e:
            logger.warning(f"OSS persistence failed for {oss_object_name}: {e}")

    async def close(self, timeout: float = 5.0) -> None:
        """Wait for pending persistence tasks (with timeout)."""
        if not self._pending_persistence_tasks:
            return
        try:
            await asyncio.wait_for(
                asyncio.gather(*self._pending_persistence_tasks, return_exceptions=True),
                timeout=timeout,
            )
        except asyncio.TimeoutError:
            logger.warning(
                f"OSS persistence tasks did not finish within {timeout}s on close "
                f"({len(self._pending_persistence_tasks)} pending)"
            )

File: rock/ts-sdk/src/sandbox/oss_client.ts
/**
 * OssClient — encapsulates all OSS interactions for a Sandbox.
 *
 * Holds OSS state (bucket, token expiration, async persistence tasks) and
 * exposes upload / download / persistence operations. Composed by Sandbox.
 *
 * Design mirrors Python SDK's rock/sdk/sandbox/oss_client.py.
 */

import { createHash } from 'crypto';
import { basename } from 'path';
import { initLogger } from '../logger.js';
import { envVars } from '../env_vars.js';
import type {
  UploadResponse,
  DownloadFileResponse,
  OssCredentials,
} from '../types/responses.js';
import { OssCredentialsSchema } from '../types/responses.js';
import type { ProgressInfo } from '../types/requests.js';

const logger = initLogger('rock.sandbox.oss');

// ─── Types ────────────────────────────────────────────────────────

/**
 * Resolved OSS configuration (Layer 1 env or Layer 2 server).
 */
export interface OssClientConfig {
  endpoint: string;
  bucket: string;
  region: string;
  enabledViaEnv: boolean; // true = Layer 1 (gated by ROCK_OSS_ENABLE); false = Layer 2
  prefix: string;
}

// ─── OssClient ─────────────────────────────────────────────────────

export class OssClient {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private bucket: any = null;
  private tokenExpireTime: string | null = null;
  private clientConfig: OssClientConfig | null = null;
  private pendingPersistenceTasks: Set<Promise<void>> = new Set();

  // We need the sandbox for API calls (get_token, arun, execute).
  // Use unknown to avoid circular import; the actual type is Sandbox from client.ts
  private sandbox: unknown;

  constructor(sandbox: unknown) {
    this.sandbox = sandbox;
  }

  // ─── Static helpers ─────────────────────────────────────────────

  /**
   * Compute deterministic OSS object name.
   *
   * Uses SHA-256(sandbox_id|local_path|sandbox_path) as digest.
   * Filename derived from sandbox_path basename (preferred) or local_path basename.
   */
  static computeObjectName(
    sandboxId: string,
    localPath: string,
    sandboxPath: string,
    prefix?: string,
  ): string {
    const payload = `${sandboxId}|${localPath}|${sandboxPath}`;
    const digest = createHash('sha256').update(payload, 'utf-8').digest('hex');
    const filename = basename(sandboxPath) || basename(localPath);

    const cleanPrefix = (prefix ?? '')
      .trim()
      .replace(/^\/+|\/+$/g, '')
      .replace(/\/+/g, '/');
    if (cleanPrefix) {
      return `${cleanPrefix}/${digest}-${filename}`;
    }
    return `${digest}-${filename}`;
  }

  /**
   * Resolve OSS configuration from two layers:
   *   Layer 1: environment variables (highest priority)
   *   Layer 2: server /get_token response (fallback)
   *   Layer 3: OSS unavailable (returns null)
   *
   * @param stsResponse - The full result dict from /get_token?account=primary
   * @returns Resolved config or null if OSS is unavailable
   */
  static resolveConfig(stsResponse: Record<string, unknown>): OssClientConfig | null {
    // Layer 1: env var (highest priority)
    const envEndpoint = process.env['ROCK_OSS_BUCKET_ENDPOINT'];
    const envBucket = process.env['ROCK_OSS_BUCKET_NAME'];
    const envRegion = process.env['ROCK_OSS_BUCKET_REGION'];
    if (envEndpoint && envBucket && envRegion) {
      return {
        endpoint: envEndpoint,
        bucket: envBucket,
        region: envRegion,
        enabledViaEnv: true,
        prefix: envVars.ROCK_OSS_TRANSFER_PREFIX ?? '',
      };
    }

    // Layer 2: server response (fallback)
    const respEndpoint =
      (stsResponse['endpoint'] as string | undefined) ??
      (stsResponse['Endpoint'] as string | undefined);
    const respBucket =
      (stsResponse['bucket'] as string | undefined) ??
      (stsResponse['Bucket'] as string | undefined);
    const respRegion =
      (stsResponse['region'] as string | undefined) ??
      (stsResponse['Region'] as string | undefined);
    const respPrefix =
      (stsResponse['prefix'] as string | undefined) ??
      (stsResponse['Prefix'] as string | undefined) ??
      '';
    if (respEndpoint && respBucket && respRegion) {
      return {
        endpoint: respEndpoint,
        bucket: respBucket,
        region: respRegion,
        enabledViaEnv: false,
        prefix: respPrefix,
      };
    }

    // Layer 3: OSS unavailable
    return null;
  }

  /**
   * Normalize an OSS region string for ali-oss / ossutil.
   */
  static normalizeRegion(region: string): string {
    return region.replace(/^oss-/, '');
  }

  // ─── Public API ─────────────────────────────────────────────────

  /**
   * Whether OSS is available: bucket has been successfully initialized.
   */
  isAvailable(): boolean {
    return this.bucket !== null;
  }

  /**
   * Ensure OSS bucket is set up and token is fresh. Idempotent.
   *
   * Returns true if OSS is available, false otherwise.
   */
  async ensureSetup(): Promise<boolean> {
    if (this.bucket !== null && !this.isTokenExpired()) {
      return true;
    }
    return this.setup();
  }

  /**
   * Upload a local file to sandbox via OSS as intermediary (large file path).
   *
   * @param timeout - Optional timeout in milliseconds
   * @param onProgress - Optional progress callback
   */
  async uploadViaOss(
    sourcePath: string,
    targetPath: string,
    timeout?: number,
    onProgress?: (info: ProgressInfo) => void,
  ): Promise<UploadResponse> {
    if (!this.bucket) {
      return { success: false, message: 'OSS bucket not set up', fileName: '' };
    }

    const sandbox = this.sandbox as { sandboxId: string; arun: Function; execute: Function; url: string; buildHeaders: Function };
    const { basename: pathBasename } = await import('path');
    const fileName = pathBasename(sourcePath);

    const ossObjectName = OssClient.computeObjectName(
      sandbox.sandboxId,
      sourcePath,
      targetPath,
      this.clientConfig?.prefix,
    );

    try {
      // Upload local file to OSS
      const { stat } = await import('fs/promises');
      const stats = await stat(sourcePath);
      const fileSize = stats.size;
      const multipartThreshold = 1024 * 1024; // 1MB

      const ossTimeout = timeout ?? envVars.ROCK_OSS_TIMEOUT;

      if (fileSize >= multipartThreshold) {
        await this.bucket.multipartUpload(ossObjectName, sourcePath, {
          timeout: ossTimeout,
          partSize: multipartThreshold,
          progress: (p: number) => {
            onProgress?.({
              phase: 'upload-to-oss',
              percent: Math.round(p * 100),
            });
          },
        });
      } else {
        await this.bucket.put(ossObjectName, sourcePath, {
          timeout: ossTimeout,
          progress: (p: number) => {
            onProgress?.({
              phase: 'upload-to-oss',
              percent: Math.round(p * 100),
            });
          },
        });
      }

      // Generate signed URL for sandbox to download
      const signedUrl = this.bucket.signatureUrl(ossObjectName, { expires: 600 });

      // Notify: starting sandbox download phase
      onProgress?.({ phase: 'download-to-sandbox', percent: -1 });

      // mkdir -p target parent
      const parentDir = targetPath.replace(/\/[^/]*$/, '').replace(/^$/, '/');
      await sandbox.arun(`mkdir -p '${parentDir}'`, { mode: 'normal', timeout: 10 });

      // wget the signed URL (nohup: avoids server-side 85s session timeout)
      const downloadCmd = `wget -O '${targetPath}' '${signedUrl}'`;
      await sandbox.arun(downloadCmd, { mode: 'nohup', waitTimeout: 600 });

      // Nohup only confirms that wget exited, not that it succeeded. Verify the
      // downloaded file is complete before reporting the upload as successful.
      const sizeCheck = await sandbox.execute({ command: ['stat', '-c', '%s', targetPath], timeout: 60 });
      const { exitCode, stdout } = sizeCheck as { exitCode?: number; stdout: string };
      const remoteSizeOutput = stdout.trim();
      if (exitCode !== 0 || !/^\d+$/.test(remoteSizeOutput)) {
        return {
          success: false,
          message: `Failed to upload file ${fileName}, unable to verify sandbox file size`,
          fileName,
        };
      }

      const remoteFileSize = Number(remoteSizeOutput);
      if (remoteFileSize !== fileSize) {
        return {
          success: false,
          message: `Failed to upload file ${fileName}, expected ${fileSize} bytes, got ${remoteFileSize} bytes`,
          fileName,
        };
      }

      return {
        success: true,
        message: `Successfully uploaded file ${fileName} to ${targetPath}`,
        fileName,
      };
    } catch (e) {
      logger.warn(`upload_via_oss failed: ${e}`);
      return {
        success: false,
        message: `Failed to upload file ${fileName} to ${targetPath}: ${e}`,
        fileName,
      };
    }
  }

  /**
   * Download file from sandbox to local via OSS as intermediary.
   *
   * Note: ensureSetup must succeed before calling. Caller (LinuxFileSystem)
   * is responsible for running ensure_ossutil in the sandbox before calling here.
   *
   * @param timeout - Optional timeout in milliseconds
   * @param onProgress - Optional progress callback
   */
  async downloadViaOss(
    remotePath: string,
    localPath: string,
    timeout?: number,
    onProgress?: (info: ProgressInfo) => void,
  ): Promise<DownloadFileResponse> {
    if (!this.bucket || !this.clientConfig) {
      return { success: false, message: 'OSS is not available' };
    }

    const sandbox = this.sandbox as { sandboxId: string; arun: Function; execute: Function; getOssStsCredentials?: Function };

    // Verify source file exists in sandbox
    const check = await sandbox.execute({ command: ['test', '-f', remotePath], timeout: 60 });
    if ((check as { exitCode: number }).exitCode !== 0) {
      return {
        success: false,
        message: `Source file not found or is not a regular file in sandbox: ${remotePath}. Note: Only regular files are supported. For directories, create a tar archive first.`,
      };
    }

    // Refresh STS creds if expired
    if (this.isTokenExpired()) {
      await this.setup();
    }
    const credentials = await this.getStsCredentials();

    // Upload sandbox file to OSS via ossutil
    const ossObjectName = OssClient.computeObjectName(
      sandbox.sandboxId,
      localPath,
      remotePath,
      this.clientConfig.prefix,
    );
    const ossUrl = `oss://${this.clientConfig.bucket}/${ossObjectName}`;
    const normalizedRegion = OssClient.normalizeRegion(this.clientConfig.region ?? '');

    const ossutilInner = `ossutil cp '${remotePath}' '${ossUrl}' --access-key-id '${credentials.accessKeyId}' --access-key-secret '${credentials.accessKeySecret}' --sts-token '${credentials.securityToken}' --endpoint '${this.clientConfig.endpoint}' --region '${normalizedRegion}'`;
    const uploadCmd = `bash -c '${ossutilInner.replace(/'/g, "'\"'\"'")}'`;
    const uploadResp = await sandbox.arun(uploadCmd, { mode: 'nohup', waitTimeout: 600 });
    if ((uploadResp as { exitCode: number }).exitCode !== 0) {
      return {
        success: false,
        message: `Failed to upload file to OSS (exit_code=${(uploadResp as { exitCode: number }).exitCode}): ${(uploadResp as { output: string }).output}`,
      };
    }

    // Download from OSS to local via ali-oss
    const { dirname: pathDirname } = (await import('path'));
    const local = typeof localPath === 'string'
      ? localPath.replace(/^~/, (process.env['HOME'] ?? '/root'))
      : localPath;
    const localDir = pathDirname(local);
    const { mkdir } = await import('fs/promises');
    await mkdir(localDir, { recursive: true });

    const ossTimeout = timeout ?? envVars.ROCK_OSS_TIMEOUT;

    try {
      await this.bucket.get(ossObjectName, local, {
        timeout: ossTimeout,
        progress: (p: number) => {
          onProgress?.({
            phase: 'download-to-local',
            percent: Math.round(p * 100),
          });
        },
      });
    } catch (e) {
      return { success: false, message: `Failed to download from OSS: ${e}` };
    }

    // Cleanup OSS object
    try {
      await this.bucket.delete(ossObjectName);
    } catch {
      // Ignore cleanup errors
    }

    return { success: true, message: `Successfully downloaded ${remotePath} to ${localPath}` };
  }

  /**
   * Fire-and-forget: upload local file to OSS in the background.
   *
   * Returns the OSS object key if scheduled, null if OSS is unavailable.
   * Failures are logged as warnings; main flow is unaffected.
   */
  async scheduleAsyncPersist(localPath: string, sandboxPath: string): Promise<string | null> {
    if (!this.bucket) {
      return null;
    }

    const sandbox = this.sandbox as { sandboxId: string };

    const ossObjectName = OssClient.computeObjectName(
      sandbox.sandboxId,
      localPath,
      sandboxPath,
      this.clientConfig?.prefix,
    );

    const task = this.persistToOss(localPath, ossObjectName)
      .then(() => {
        logger.info(`OSS persisted: ${ossObjectName}`);
      })
      .catch((e: unknown) => {
        logger.warn(`OSS persistence failed for ${ossObjectName}: ${e}`);
      });

    this.pendingPersistenceTasks.add(task);
    const untrack = () => { this.pendingPersistenceTasks.delete(task); };
    task.then(untrack).catch(untrack);

    return ossObjectName;
  }

  /**
   * Wait for pending persistence tasks (with timeout).
   *
   * @param timeoutMs - Maximum time to wait in milliseconds (default 5000)
   */
  async close(timeoutMs: number = 5000): Promise<void> {
    if (this.pendingPersistenceTasks.size === 0) {
      return;
    }

    const all = Promise.allSettled([...this.pendingPersistenceTasks]);
    const timeout = new Promise<void>((resolve) => setTimeout(resolve, timeoutMs));
    await Promise.race([all, timeout]);

    if (this.pendingPersistenceTasks.size > 0) {
      logger.warn(
        `OSS persistence tasks did not finish within ${timeoutMs}ms on close ` +
        `(${this.pendingPersistenceTasks.size} pending)`,
      );
    }
  }

  // ─── Private helpers ────────────────────────────────────────────

  /**
   * Fetch STS credentials and OSS config from /get_token endpoint.
   */
  private async getStsCredentials(): Promise<OssCredentials> {
    const sandbox = this.sandbox as { url: string; buildHeaders: () => Record<string, string> };
    const HttpUtils = (await import('../utils/http.js')).HttpUtils;

    const url = `${sandbox.url}/get_token?account=primary`;
    const headers = sandbox.buildHeaders();
    const response = await HttpUtils.get<OssCredentials>(url, headers);

    if (response.status !== 'Success') {
      throw new Error(`Failed to get OSS STS token: ${JSON.stringify(response)}`);
    }

    const credentials = OssCredentialsSchema.parse(response.result);
    this.tokenExpireTime = credentials.expiration;
    return credentials;
  }

  /**
   * Whether cached token is missing, malformed, or within 5min of expiration.
   */
  private isTokenExpired(): boolean {
    if (!this.tokenExpireTime) {
      return true;
    }

    try {
      const expireTime = new Date(this.tokenExpireTime);
      const currentTime = new Date();
      const bufferMs = 5 * 60 * 1000; // 5 minutes
      return currentTime.getTime() >= (expireTime.getTime() - bufferMs);
    } catch {
      return true;
    }
  }

  /**
   * Core setup: get STS credentials -> resolve config -> initialize ali-oss bucket.
   *
   * Returns true if OSS is successfully set up, false otherwise.
   */
  private async setup(): Promise<boolean> {
    let credentials: OssCredentials;
    try {
      credentials = await this.getStsCredentials();
    } catch (e) {
      logger.warn(`Failed to get STS credentials: ${e}`);
      return false;
    }

    const config = OssClient.resolveConfig(credentials as unknown as Record<string, unknown>);
    if (!config) {
      return false;
    }

    // Layer 1 also requires ROCK_OSS_ENABLE
    if (
      config.enabledViaEnv &&
      process.env['ROCK_OSS_ENABLE']?.toLowerCase() !== 'true'
    ) {
      return false;
    }

    try {
      const OSS = (await import('ali-oss')).default;

      const ossTimeout = envVars.ROCK_OSS_TIMEOUT;

      this.bucket = new OSS({
        secure: true,
        timeout: ossTimeout,
        endpoint: config.endpoint,
        region: OssClient.normalizeRegion(config.region),
        accessKeyId: credentials.accessKeyId,
        accessKeySecret: credentials.accessKeySecret,
        stsToken: credentials.securityToken,
        bucket: config.bucket,
        refreshSTSToken: async () => {
          const newCreds = await this.getStsCredentials();
          return {
            accessKeyId: newCreds.accessKeyId,
            accessKeySecret: newCreds.accessKeySecret,
            stsToken: newCreds.securityToken,
          };
        },
        refreshSTSTokenInterval: 300000, // 5 minutes
      });

      this.clientConfig = config;
      return true;
    } catch (e) {
      logger.warn(`Failed to initialize OSS bucket: ${e}`);
      this.bucket = null;
      return false;
    }
  }

  /**
   * Execute a single persistence upload to OSS.
   */
  private async persistToOss(localPath: string, ossObjectName: string): Promise<void> {
    const { stat } = await import('fs/promises');
    const stats = await stat(localPath);
    const fileSize = stats.size;
    const multipartThreshold = 1024 * 1024; // 1MB

    if (fileSize >= multipartThreshold) {
      await this.bucket.multipartUpload(ossObjectName, localPath, {
        partSize: multipartThreshold,
        timeout: envVars.ROCK_OSS_TIMEOUT,
      });
    } else {
      await this.bucket.put(ossObjectName, localPath, {
        timeout: envVars.ROCK_OSS_TIMEOUT,
      });
    }
  }
}
